From 9fed05a4abd0314274e80cc4755acdcef6c7eb5e Mon Sep 17 00:00:00 2001
From: Mohit Setia <msetia@codeaurora.org>
Date: Mon, 16 Sep 2013 16:05:48 +0530
Subject: [PATCH 018/176] QCamera2 :  memory leak with memory buffer
 allocations and deallocation

native_handle_t struct allocated not freed in deallocation is fixed.

Change-Id: I136c15470a2ccadbd5267a0d888e58eb147f5460
---
 QCamera2/HAL/QCameraMem.cpp | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/QCamera2/HAL/QCameraMem.cpp b/QCamera2/HAL/QCameraMem.cpp
index 8dcb62d..76cc9a3 100644
--- a/QCamera2/HAL/QCameraMem.cpp
+++ b/QCamera2/HAL/QCameraMem.cpp
@@ -950,6 +950,22 @@ int QCameraVideoMemory::allocateMore(int count, int size)
 void QCameraVideoMemory::deallocate()
 {
     for (int i = 0; i < mBufferCount; i ++) {
+        struct encoder_media_buffer_type * packet =
+            (struct encoder_media_buffer_type *)mMetadata[i]->data;
+        if (NULL != packet) {
+            native_handle_t * nh = const_cast<native_handle_t *>(packet->meta_handle);
+            if (NULL != nh) {
+               if (native_handle_delete(nh)) {
+                   ALOGE("Unable to delete native handle");
+               }
+            }
+            else {
+               ALOGE("native handle not available");
+            }
+        }
+        else {
+            ALOGE("packet not available");
+        }
         mMetadata[i]->release(mMetadata[i]);
         mMetadata[i] = NULL;
     }
-- 
1.8.3.1

